### ISSUE TITLE
Determine best G1GC flags to avoid too much compromise on performance and memory consumption

### ISSUE THREAD

**Issue by andsel** [elastic/logstash#13378]
This is issue tracks the effort to investigate the best G1GC options:

G1 has the statement to minimize GC pauses at the expense of throughput. This issue is a notebook used to track discoveries.

**First basic experiment**
Pipeline used:
```
input {
  generator {
    message => '{"some": "value"}'
  }
}

output {
  sink {
    enable_metric => "true"
}
```

Tracking GC cpu time and consumed memory between CMS (JDK 11) and G1(JDK 11) becomes evident that while the CPU % remain the same the Java heap fluctuate more on G1.(screenshot from JVisualVM)

**CMS JDK 11**
![CMS_jdk11](https://user-images.githubusercontent.com/1038545/139862540-9274c3a1-f05f-4e51-bdb7-12105044a636.png)

**G1 JDK 17**
![G1_jdk17](https://user-images.githubusercontent.com/1038545/139862661-6935597e-20c5-45bc-95f3-41428cdca9fe.png)

**Comment by kares** [comment]
:+1: great work, the `500ms` is tempting but I am not convinced it should be set (but also have no objections against).

was wondering if reducing `-XX:MaxTenuringThreshold=8` would reduce the increased CPU amount you've seen.
most garbage is short lived (event data) thus the default (15) might be a too high for typical LS use-cases and since G1 also compacts opposed to CMS we might want to help it at least reduce some time spent copying objects around.

some low hanging fruits if we're reviewing and updating *config/jvm.options* anyway:

1. let's set meta-space size explicitly based on size we're seeing in Logstash (~85M) e.g. `-XX:MetaspaceSize=100M`
    should avoid early Full GCs

2. `-XX:+ParallelRefProcEnabled` - parallelize reference processing (enabled since Java 17)

3. `-XX:+AlwaysPreTouch` - helps avoid some latency at runtime

**Comment by andsel** [comment]
> 1. let's set meta-space size explicitly based on size we're seeing in Logstash .. should avoid early Full GCs

No, because those FullGCs come from explicit invokcation of `System.gc()` maybe could be avoided with the JVM flag `-XX:+DisableExplicitGC`

> 2. `-XX:+ParallelRefProcEnabled` - parallelize reference processing (enabled since Java 17)

Tested but as with following image no appreciable improvements, maybe a little loss. On left the flag is disabled on right is enabled
![g1_without_and_with_ParallelRefProcEnabled](https://user-images.githubusercontent.com/1038545/141335527-dca019b2-1767-42e8-a4a5-36be7e534ae0.png)

> 3. `-XX:+AlwaysPreTouch` - helps avoid some latency at runtime

Same here, no improvements (at least qualitatively), on left the flag is disable, on right is enabled
![g1_without_vs_with_AlwaysPreTouch](https://user-images.githubusercontent.com/1038545/141335698-08bf0b9b-92e8-4620-a678-259b7fc77175.png)

**Comment by kares** [comment]
yeah those low hanging flags weren't mean to make a difference - just to check everything is the same as before ...

even reducing `-XX:MaxTenuringThreshold=8` might not show much but would be interesting to see if CPU usage drops a bit.

**Comment by andsel** [comment]
Hi @kares tried with `-XX:MaxTenuringThreshold=8`  (on the right) but CPU consumption keeps the same while throughput is lower about ~2K evt/sec
![g1_without_and_withTenuringThreshold8](https://user-images.githubusercontent.com/1038545/142391846-1a7c29c3-2aa1-4b2b-bfde-e0c1c86c1d72.png)

**Comment by jsvd** [comment]
Here are my results: https://gist.github.com/jsvd/4ccbc1776f3bc1679e4751aa0a9500be

Key takeaways:
1. in a medium complexity pipeline, 1gb heaps hinder throughput and latency when compared to 4gb heaps;
2. throughput-wise, with 4gb heaps size, Java 17 w/ G1 performs better than Java 11 w/CMS;
3. Regarding G1, tweaks to flags such as _MaxGCPauseMillis, InitiatingHeapOccupancyPercent_ and _G1ReservePercent_ didn't show significant improvement in performance or latency.
4. Process wise, we don't have a reproducible tool to capture throughput and latency (including percentiles), which severely limits the ability to do benchmarking and decide on performance configuration.

**Comment by robbavey** [comment]
@andsel @jsvd Is there anything to follow up on this?

**Comment by andsel** [comment]
No, we could eventually define some configurations that we can use to benchmark Logstash in different uses cases, so that we can spend less time in thinking how to test next time we come down to the same theme.
